Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+    }
+
+    if (to_parse.is_empty()) {
+        return { command, end_of_command_index };
+    }
+
+    char command_char = to_parse[0];
+    to_parse = to_parse.substring_view(1);
+    end_of_command_index += 1;
+#define HANDLE_FUNCTION_CASE(c, max_addresses)                     \
+    case c:                                                        \
+        command.function = c;                                      \
+        if (!verify_number_of_addresses<max_addresses>(command)) { \
+            return { {}, end_of_command_index };                   \
+        }                                                          \
+        break;
+    switch (command_char) {
+        ENUMERATE_FUNCTIONS(HANDLE_FUNCTION_CASE)
+    default:
+        Core::File::standard_error()->write(String::formatted("Unknown function command '{}'\n", to_parse[0]));
+        return { command, end_of_command_index };
+    }
+#undef HANDLE_FUNCTION_CASE
+
+    // TODO: improve argument parsing
+    if (to_parse.is_empty())
+        return { command, end_of_command_index };
+    auto command_separator_index = to_parse.find(';').value_or(to_parse.length() - 1);
+    command.arguments = to_parse.substring_view(0, command_separator_index);
+    end_of_command_index += command_separator_index + 1;
+    return { command, end_of_command_index };
+}
+
+class Script {
+public:
+    [[nodiscard]] bool add_script_part(StringView data)
+    {
+        StringView command_sv = data.trim_whitespace();
+        while (!command_sv.is_empty()) {
+            auto [command, end_of_command_index] = parse_command(command_sv);
+            if (!command.has_value())
+                return false;
+            m_commands.append(command.release_value());
+            command_sv = command_sv.substring_view(end_of_command_index).trim_whitespace();
+        }
+        return true;
+    }
+
+    Vector<Command>& commands() { return m_commands; }
+
+private:
+    Vector<Command> m_commands;
+};
+
+enum class CycleDecision {
+    None,
+    Next,
+    Quit
+};
+
+class InputFile {
+    AK_MAKE_NONCOPYABLE(InputFile);
+
+public:
+    InputFile(RefPtr<Core::File> file)
+        : m_file(move(file))
+        , m_line_iterator(m_file->line_begin())
+    {
+    }
+    InputFile(InputFile&&) = default;
+    InputFile& operator=(InputFile&&) = default;
+
+    bool has_next() const
+    {
+        return !m_line_iterator.at_end();
+    }
+
+    StringView next()
+    {
+        VERIFY(has_next());
+        if (!m_started)
+            m_started = true;
+        else
+            ++m_line_iterator;
+        auto line = *m_line_iterator;
+        ++m_line_number;
+        return line;
+    }
+
+    size_t line_number() const { return m_line_number; }
+
+private:
+    RefPtr<Core::File> m_file;
+    Core::LineIterator m_line_iterator;
+    size_t m_line_number { 0 };
+    bool m_started { false };
+};
+
+static void write_pattern_space(RefPtr<Core::File> output, StringBuilder& pattern_space)
```

Comment:
Can this take a `Core::File&` instead? no need to go through a ref/unref just to treat it as a simple reference.

Reply:
Sure can, fixed.now